sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import scala.collection.mutable
+
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.jni.{RetryOOM, RmmSpark, SplitAndRetryOOM}
+
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+object RmmRapidsRetryIterator extends Arm {
+  def splitInHalfByRows(
+      spillable: SpillableColumnarBatch): Seq[SpillableColumnarBatch] = {
+    val toSplitRows = spillable.numRows
+    if (toSplitRows <= 1) {
+      throw new OutOfMemoryError(s"A batch of $toSplitRows cannot be split!")
+    }
+    val (firstHalf, secondHalf) = withResource(spillable.getColumnarBatch()) { src =>
+      withResource(GpuColumnVector.from(src)) { tbl =>
+        val splitIx = (tbl.getRowCount / 2).toInt
+        withResource(tbl.contiguousSplit(splitIx)) { cts =>
+          val batches = cts.map(_.getTable).map(GpuColumnVector.from(_, spillable.dataTypes))
+          val spillables = batches.map { b =>
+            SpillableColumnarBatch(
+              b,
+              SpillPriorities.ACTIVE_BATCHING_PRIORITY,
+              // TODO: need to figure out how to pick the right callback
+              RapidsBuffer.defaultSpillCallback)
+          }
+          require(spillables.length == 2)
+          (spillables.head, spillables.last)
+        }
+      }
+    }
+    Seq(firstHalf, secondHalf)
+  }
+
+  private def toSpillableIterator(
+      it: Iterator[ColumnarBatch]): Iterator[SpillableColumnarBatch] = {
+    new Iterator[SpillableColumnarBatch] {
+      override def hasNext: Boolean = it.hasNext
+      override def next(): SpillableColumnarBatch = {
+        SpillableColumnarBatch(
+          it.next(),
+          SpillPriorities.ACTIVE_ON_DECK_PRIORITY,
+          RapidsBuffer.defaultSpillCallback)
```

Comment:
Again we need some kind of a better spill callback. At a minimum we should have a TODO here